File: demos/go/basic_proxy.go
// Outbound LB - Go Demo
//
// Demonstrates how to use Outbound LB proxy with Go using net/http.
// Includes examples for HTTP, HTTPS, authentication, error handling, and concurrent requests.
//
// Usage:
//
//	go run basic_proxy.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"sync"
	"time"
)

// Configuration from environment variables
var (
	proxyHost = getEnv("PROXY_HOST", "localhost")
	proxyPort = getEnv("PROXY_PORT", "3128")
	proxyUser = getEnv("PROXY_USER", "user")
	proxyPass = getEnv("PROXY_PASS", "password")
)

func getEnv(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}

func getProxyURL() *url.URL {
	proxyURL, _ := url.Parse(fmt.Sprintf("http://%s:%s", proxyHost, proxyPort))
	return proxyURL
}

func getProxyURLWithAuth() *url.URL {
	proxyURL, _ := url.Parse(fmt.Sprintf("http://%s:%s@%s:%s", proxyUser, proxyPass, proxyHost, proxyPort))
	return proxyURL
}

// IPResponse represents the response from httpbin.org/ip
type IPResponse struct {
	Origin string `json:"origin"`
}

func printSeparator(title string) {
	fmt.Println(string(make([]byte, 60)))
	fmt.Printf("%s\n", "============================================================")
	fmt.Printf("%s\n", title)
	fmt.Printf("%s\n", "============================================================")
}

// Example 1: Basic HTTP Request
func exampleHTTPRequest() {
	printSeparator("Example 1: Basic HTTP Request")

	client := &http.Client{
		Transport: &http.Transport{
			Proxy: http.ProxyURL(getProxyURL()),
		},
		Timeout: 10 * time.Second,
	}

	resp, err := client.Get("http://httpbin.org/ip")
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		fmt.Println()
		return
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	fmt.Printf("Status: %d\n", resp.StatusCode)
	fmt.Printf("Response: %s\n", body)
	fmt.Println()
}

// Example 2: HTTPS Request (CONNECT tunnel)
func exampleHTTPSRequest() {
	printSeparator("Example 2: HTTPS Request (CONNECT tunnel)")

	client := &http.Client{
		Transport: &http.Transport{
			Proxy: http.ProxyURL(getProxyURL()),
		},
		Timeout: 10 * time.Second,
	}

	resp, err := client.Get("https://httpbin.org/ip")
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		fmt.Println()
		return
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	fmt.Printf("Status: %d\n", resp.StatusCode)
	fmt.Printf("Response: %s\n", body)
	fmt.Println()
}

// Example 3: Authenticated Proxy
func exampleAuthenticatedProxy() {
	printSeparator("Example 3: Authenticated Proxy")

	client := &http.Client{
		Transport: &http.Transport{
			Proxy: http.ProxyURL(getProxyURLWithAuth()),
		},
		Timeout: 10 * time.Second,
	}

	resp, err := client.Get("https://httpbin.org/ip")
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		fmt.Println()
		return
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	fmt.Printf("Status: %d\n", resp.StatusCode)
	fmt.Printf("Response: %s\n", body)
	fmt.Println()
}

// Example 4: Error Handling
func exampleErrorHandling() {
	printSeparator("Example 4: Error Handling")

	// Test connection timeout
	fmt.Println("Testing connection timeout...")
	client := &http.Client{
		Transport: &http.Transport{
			Proxy: http.ProxyURL(getProxyURL()),
		},
		Timeout: 2 * time.Second,
	}

	_, err := client.Get("http://httpbin.org/delay/5")
	if err != nil {
		fmt.Println("  Caught timeout error (expected)")
	}

	// Test invalid URL
	fmt.Println("Testing invalid URL...")
	client2 := &http.Client{
		Transport: &http.Transport{
			Proxy: http.ProxyURL(getProxyURL()),
		},
		Timeout: 5 * time.Second,
	}

	_, err = client2.Get("http://invalid.invalid.invalid")
	if err != nil {
		fmt.Printf("  Caught connection error (expected): %T\n", err)
	}

	// Test proxy authentication failure
	fmt.Println("Testing proxy auth failure (if auth required on proxy)...")
	badProxyURL, _ := url.Parse(fmt.Sprintf("http://wronguser:wrongpass@%s:%s", proxyHost, proxyPort))
	client3 := &http.Client{
		Transport: &http.Transport{
			Proxy: http.ProxyURL(badProxyURL),
		},
		Timeout: 5 * time.Second,
	}

	resp, err := client3.Get("http://httpbin.org/ip")
	if err != nil {
		fmt.Printf("  Error: %v\n", err)
	} else {
		defer resp.Body.Close()
		if resp.StatusCode == 407 {
			fmt.Println("  Got 407 Proxy Authentication Required (expected)")
		} else {
			fmt.Printf("  Status: %d\n", resp.StatusCode)
		}
	}

	fmt.Println()
}

// Example 5: Concurrent Requests (Load Balancing Demo)
func exampleConcurrentRequests() {
	printSeparator("Example 5: Concurrent Requests (Load Balancing Demo)")

	numRequests := 10
	fmt.Printf("Making %d concurrent requests...\n", numRequests)

	type result struct {
		requestID int
		ip        string
		err       error
	}

	results := make(chan result, numRequests)
	var wg sync.WaitGroup

	client := &http.Client{
		Transport: &http.Transport{
			Proxy:               http.ProxyURL(getProxyURL()),
			MaxIdleConns:        100,
			MaxIdleConnsPerHost: 100,
		},
		Timeout: 10 * time.Second,
	}

	for i := 0; i < numRequests; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()

			resp, err := client.Get("http://httpbin.org/ip")
			if err != nil {
				results <- result{requestID: id, err: err}
				return
			}
			defer resp.Body.Close()

			var ipResp IPResponse
			if err := json.NewDecoder(resp.Body).Decode(&ipResp); err != nil {
				results <- result{requestID: id, err: err}
				return
			}

			results <- result{requestID: id, ip: ipResp.Origin}
		}(i)
	}

	// Close results channel when all goroutines complete
	go func() {
		wg.Wait()
		close(results)
	}()

	// Collect results
	ipCounts := make(map[string]int)
	for r := range results {
		if r.err != nil {
			fmt.Printf("  Request %d: error: %v\n", r.requestID, r.err)
		} else {
			fmt.Printf("  Request %d: %s\n", r.requestID, r.ip)
			ipCounts[r.ip]++
		}
	}

	fmt.Println("\nIP Distribution:")
	for ip, count := range ipCounts {
		fmt.Printf("  %s: %d requests\n", ip, count)
	}

	fmt.Println()
}

// Example 6: Context with Cancellation
func exampleContextCancellation() {
	printSeparator("Example 6: Context with Cancellation")

	client := &http.Client{
		Transport: &http.Transport{
			Proxy: http.ProxyURL(getProxyURL()),
		},
	}

	// Create a context with timeout
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, "GET", "https://httpbin.org/ip", nil)
	if err != nil {
		fmt.Printf("Error creating request: %v\n", err)
		fmt.Println()
		return
	}

	resp, err := client.Do(req)
	if err != nil {
		fmt.Printf("Error: %v\n", err)
		fmt.Println()
		return
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)
	fmt.Printf("Status: %d\n", resp.StatusCode)
	fmt.Printf("Response: %s\n", body)
	fmt.Println()
}

func main() {
	fmt.Println()
	fmt.Println("Outbound LB - Go Demo")
	fmt.Printf("Proxy: http://%s:%s\n", proxyHost, proxyPort)
	fmt.Println()

	exampleHTTPRequest()
	exampleHTTPSRequest()
	exampleAuthenticatedProxy()
	exampleErrorHandling()
	exampleConcurrentRequests()
	exampleContextCancellation()

	fmt.Println("All examples completed!")
}

File: demos/java/BasicProxy.java
/**
 * Outbound LB - Java Demo
 *
 * Demonstrates how to use Outbound LB proxy with Java using HttpClient.
 * Includes examples for HTTP, HTTPS, authentication, error handling, and concurrent requests.
 *
 * Requirements:
 *     Java 11+
 *
 * Usage:
 *     javac BasicProxy.java
 *     java BasicProxy
 */

import java.net.Authenticator;
import java.net.InetSocketAddress;
import java.net.PasswordAuthentication;
import java.net.ProxySelector;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpTimeoutException;
import java.time.Duration;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class BasicProxy {

    // Configuration from environment variables
    private static final String PROXY_HOST = System.getenv().getOrDefault("PROXY_HOST", "localhost");
    private static final int PROXY_PORT = Integer.parseInt(System.getenv().getOrDefault("PROXY_PORT", "3128"));
    private static final String PROXY_USER = System.getenv().getOrDefault("PROXY_USER", "user");
    private static final String PROXY_PASS = System.getenv().getOrDefault("PROXY_PASS", "password");

    public static void main(String[] args) {
        System.out.println();
        System.out.println("Outbound LB - Java Demo");
        System.out.printf("Proxy: http://%s:%d%n", PROXY_HOST, PROXY_PORT);
        System.out.println();

        exampleHttpRequest();
        exampleHttpsRequest();
        exampleAuthenticatedProxy();
        exampleErrorHandling();
        exampleConcurrentRequests();

        System.out.println("All examples completed!");
    }

    private static void printSeparator(String title) {
        System.out.println("=".repeat(60));
        System.out.println(title);
        System.out.println("=".repeat(60));
    }

    /**
     * Example 1: Basic HTTP Request
     */
    private static void exampleHttpRequest() {
        printSeparator("Example 1: Basic HTTP Request");

        try {
            HttpClient client = HttpClient.newBuilder()
                    .proxy(ProxySelector.of(new InetSocketAddress(PROXY_HOST, PROXY_PORT)))
                    .connectTimeout(Duration.ofSeconds(10))
                    .build();

            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create("http://httpbin.org/ip"))
                    .timeout(Duration.ofSeconds(10))
                    .GET()
                    .build();

            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

            System.out.println("Status: " + response.statusCode());
            System.out.println("Response: " + response.body());
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }

        System.out.println();
    }

    /**
     * Example 2: HTTPS Request (CONNECT tunnel)
     */
    private static void exampleHttpsRequest() {
        printSeparator("Example 2: HTTPS Request (CONNECT tunnel)");

        try {
            HttpClient client = HttpClient.newBuilder()
                    .proxy(ProxySelector.of(new InetSocketAddress(PROXY_HOST, PROXY_PORT)))
                    .connectTimeout(Duration.ofSeconds(10))
                    .build();

            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create("https://httpbin.org/ip"))
                    .timeout(Duration.ofSeconds(10))
                    .GET()
                    .build();

            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

            System.out.println("Status: " + response.statusCode());
            System.out.println("Response: " + response.body());
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }

        System.out.println();
    }

    /**
     * Example 3: Authenticated Proxy
     */
    private static void exampleAuthenticatedProxy() {
        printSeparator("Example 3: Authenticated Proxy");

        try {
            // Create an authenticator for proxy auth
            Authenticator authenticator = new Authenticator() {
                @Override
                protected PasswordAuthentication getPasswordAuthentication() {
                    if (getRequestorType() == RequestorType.PROXY) {
                        return new PasswordAuthentication(PROXY_USER, PROXY_PASS.toCharArray());
                    }
                    return null;
                }
            };

            HttpClient client = HttpClient.newBuilder()
                    .proxy(ProxySelector.of(new InetSocketAddress(PROXY_HOST, PROXY_PORT)))
                    .authenticator(authenticator)
                    .connectTimeout(Duration.ofSeconds(10))
                    .build();

            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create("https://httpbin.org/ip"))
                    .timeout(Duration.ofSeconds(10))
                    .GET()
                    .build();

            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

            System.out.println("Status: " + response.statusCode());
            System.out.println("Response: " + response.body());
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage());
        }

        System.out.println();
    }

    /**
     * Example 4: Error Handling
     */
    private static void exampleErrorHandling() {
        printSeparator("Example 4: Error Handling");

        HttpClient client = HttpClient.newBuilder()
                .proxy(ProxySelector.of(new InetSocketAddress(PROXY_HOST, PROXY_PORT)))
                .connectTimeout(Duration.ofSeconds(10))
                .build();

        // Test connection timeout
        System.out.println("Testing connection timeout...");
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create("http://httpbin.org/delay/5"))
                    .timeout(Duration.ofSeconds(2)) // Short timeout
                    .GET()
                    .build();

            client.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (HttpTimeoutException e) {
            System.out.println("  Caught timeout error (expected)");
        } catch (Exception e) {
            System.out.println("  Error: " + e.getClass().getSimpleName());
        }

        // Test invalid URL
        System.out.println("Testing invalid URL...");
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create("http://invalid.invalid.invalid"))
                    .timeout(Duration.ofSeconds(5))
                    .GET()
                    .build();

            client.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (Exception e) {
            System.out.println("  Caught connection error (expected): " + e.getClass().getSimpleName());
        }

        // Test proxy auth failure
        System.out.println("Testing proxy auth failure (if auth required on proxy)...");
        try {
            // Create authenticator with wrong credentials
            Authenticator badAuthenticator = new Authenticator() {
                @Override
                protected PasswordAuthentication getPasswordAuthentication() {
                    if (getRequestorType() == RequestorType.PROXY) {
                        return new PasswordAuthentication("wronguser", "wrongpass".toCharArray());
                    }
                    return null;
                }
            };

            HttpClient badClient = HttpClient.newBuilder()
                    .proxy(ProxySelector.of(new InetSocketAddress(PROXY_HOST, PROXY_PORT)))
                    .authenticator(badAuthenticator)
                    .connectTimeout(Duration.ofSeconds(5))
                    .build();

            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create("http://httpbin.org/ip"))
                    .timeout(Duration.ofSeconds(5))
                    .GET()
                    .build();

            HttpResponse<String> response = badClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() == 407) {
                System.out.println("  Got 407 Proxy Authentication Required (expected)");
            } else {
                System.out.println("  Status: " + response.statusCode());
            }
        } catch (Exception e) {
            System.out.println("  Error: " + e.getMessage());
        }

        System.out.println();
    }

    /**
     * Example 5: Concurrent Requests (Load Balancing Demo)
     */
    private static void exampleConcurrentRequests() {
        printSeparator("Example 5: Concurrent Requests (Load Balancing Demo)");

        int numRequests = 10;
        System.out.printf("Making %d concurrent requests...%n", numRequests);

        ExecutorService executor = Executors.newFixedThreadPool(5);

        HttpClient client = HttpClient.newBuilder()
                .proxy(ProxySelector.of(new InetSocketAddress(PROXY_HOST, PROXY_PORT)))
                .connectTimeout(Duration.ofSeconds(10))
                .executor(executor)
                .build();

        List<CompletableFuture<String>> futures = new ArrayList<>();

        for (int i = 0; i < numRequests; i++) {
            final int requestId = i;

            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create("http://httpbin.org/ip"))
                    .timeout(Duration.ofSeconds(10))
                    .GET()
                    .build();

            CompletableFuture<String> future = client
                    .sendAsync(request, HttpResponse.BodyHandlers.ofString())
                    .thenApply(response -> {
                        String ip = extractOrigin(response.body());
                        System.out.printf("  Request %d: %s%n", requestId, ip);
                        return ip;
                    })
                    .exceptionally(e -> {
                        System.out.printf("  Request %d: error: %s%n", requestId, e.getMessage());
                        return "error";
                    });

            futures.add(future);
        }

        // Wait for all requests to complete
        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])).join();

        // Count IP distribution
        Map<String, Integer> ipCounts = new HashMap<>();
        for (CompletableFuture<String> future : futures) {
            try {
                String ip = future.get();
                if (!ip.equals("error")) {
                    ipCounts.merge(ip, 1, Integer::sum);
                }
            } catch (Exception ignored) {
            }
        }

        System.out.println("\nIP Distribution:");
        ipCounts.forEach((ip, count) -> System.out.printf("  %s: %d requests%n", ip, count));

        executor.shutdown();
        System.out.println();
    }

    /**
     * Extract the "origin" field from JSON response
     */
    private static String extractOrigin(String json) {
        // Simple extraction without JSON library
        int start = json.indexOf("\"origin\"");
        if (start == -1) return "unknown";

        int colon = json.indexOf(":", start);
        int quote1 = json.indexOf("\"", colon);
        int quote2 = json.indexOf("\"", quote1 + 1);

        if (quote1 != -1 && quote2 != -1) {
            return json.substring(quote1 + 1, quote2);
        }
        return "unknown";
    }
}
